### DISCUSSION TITLE
Updating BetterDiscord from 1.8.2 to 1.8.3/1.8.4

### DISCUSSION THREAD

**Discussion by tsukasa** [tsukasa/BdBrowser#12] [Announcements]
Greetings,

*Edit 2022-12-14 21:45 GMT+1: If you are already on BD 1.8.3 and want to update to 1.8.4, you can use BetterDiscord's built-in updater in the settings. The steps outlined below are only required if BetterDiscord fails to initialize.*

due to changes made by Discord, BetterDiscord fails to initialize unless it gets updated to version 1.8.3 or newer - meaning we cannot use the built-in updater to perform this task.

Since BetterDiscord refuses to load wholesale, please delete the `betterdiscord.asar` from the virtual filesystem and refresh the page to force BdBrowser into downloading the latest version from BetterDiscord's GitHub releases page:

1. Open the Developer Tools console.
2. Delete the betterdiscord.asar:
   ```javascript
    require("fs").unlinkSync("AppData/BetterDiscord/data/betterdiscord.asar");
   ```
3. Refresh the page.

Please note that as with every update Discord pushes, there is a chance that your plugins will break. These issues are beyond the scope of this project and must be addressed by the plugin authors individually.

As far as I can tell, no update of the BdBrowser extension is required.

**Comment by ClaraCrazy** [comment]
Cannot contact github "because it violates the following Content Security Policy directive: "connect-src 'self' "

**Reply by ClaraCrazy** [reply]
even when curling the new asar myself and supplying it otherwise, loading BD now fails with an error thrown at an unknown location failing to use a `toString` function.

**Reply by tsukasa** [reply]
Greetings,

the toString() message is indicative of an issue stemming from the renderer.

To ensure no odd ends are causing trouble, please check the following points:

1. If you are using a local betterdiscord.js from the extension's `dist` directory, make sure it is a build of BetterDiscord 1.8.3 - otherwise this error is to be expected as the local renderer is used in favour of the betterdiscord.asar from the VFS when present.
2. Do a hard refresh (Ctrl+F5) to prevent cached fragments from causing trouble.

The CSP error is indicative of the service worker not being present or the DNF rules set by the service worker being purged, as responses from requests of Discord's client domains automatically [lose their Content-Security-Policy header](https://github.com/tsukasa/BdBrowser/blob/master/service/src/modules/declarativenetrequest.js#L24).

You can double-check this by navigating to the Chrome extension page, clicking details on BdBrowser and clicking the "service worker" link to bring up the console. If you issue the following command, you should get back at least two rules - one of which removes the Content-Security-Policy header for discord.com, canary.discord.com and ptb.discord.com:

```javascript
await chrome.declarativeNetRequest.getSessionRules();
```

If the rule is missing, the easiest way to re-instate it would be to disable/enable the extension so that the service worker restarts.

**Reply by ClaraCrazy** [reply]
now its working again. Thanks mate. <3

**Comment by vjbenn** [comment]
hello thank you for the quick announcement 

I got a problem when I copy that script thru developer tools's console

`Uncaught Error: ENOENT: no such file or directory, unlink 'AppData/BetterDiscord/data/betterdiscord.asar'
    at getVfsErrorObject (frontend.js:1727:24)
    at Object.unlinkSync (frontend.js:2039:32)
    at <anonymous>:1:15`
    
Sorry I dont have an IT background and clearly this is skill issue. can you elaborate right step please? thank you

**Reply by tsukasa** [reply]
Greetings,

generally speaking the error message in itself occurring is okay, it means that there was no `betterdiscord.asar` present in the virtual filesystem. If you are using the latest version of BdBrowser and refresh the page, the extension should automatically discover that the `betterdiscord.asar` file is missing in the VFS and try to download it from BetterDiscord's GitHub releases. After downloading the file, the BetterDiscord renderer should initialize as you would expect. So much for the theory, at least.

You can manually verify the existence of the file by using the Developer Tools console and issuing the following command:
```javascript
require("fs").statSync("AppData/BetterDiscord/data/betterdiscord.asar");
```

If you get another `ENOENT: no such file or directory` error, this means that the download did not happen. This can have multiple possible reasons:
1. Perhaps the `data` directory does not exist in the virtual filesystem. You can double-check this through the console as well:
   ```javascript
   require("fs").statSync("AppData/BetterDiscord/data/");
   ```
   If this comes up with anything other than an error message, you should be fine and it is probably one of the other points.
   Probability: Low.
1. It is possible the download was blocked.
   In this case, you should find a console log entry complaining about the exact nature of why that happened.
   Probability: Relatively low.
3. The extension might be outdated.
   Up to November I often pushed changes within a short interval, so it is possible you are using an older version of the extension because, well... why not. The feature to download/load the `betterdiscord.asar` was a relatively late addition.
   
   If you are unsure whether or not you *do* run the latest version, you can follow the guide in the README.md file for updating the extension. The tl;dr is to pull the latest version, go to your Chrome extension page, click the "Update" button and do a hard refresh of the page to ensure old scripts and discarded service workers get replaced.
   
   A small details widget should be inserted into the Discord settings menu:
   ```
   Stable 163880 (4ef3df8)
   BDBrowser 1.8.2.20221110
   betterdiscord.asar (v1.8.3, VFS)
   Windows 10 64-Bit
   ```
   
   If it looks like that, then everything should be peachy. A BDBrowser version mismatch can be remedied with the steps explained above. If the entry for "betterdiscord" states "betterdiscord.js (vXXX, local)" instead, you should rename the `dist/betterdiscord.js` in the extension's directory to a temp name and reload the page to test if it helps.

**Reply by vjbenn** [reply]
its working. thank you so much. have a good day!
